plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/expression/Meta.kt

```diff
@@ -29,3 +35,87 @@ fun ExpEmbedding.withPosition(source: KtSourceElement?): ExpEmbedding =
         else -> WithPosition(this, source)
     }
 
+
+/**
+ * Represents a subtree in which a different subtree may appear multiple times.
+ *
+ * This is a complicated construction. There are cases when we want to use a source-level construction multiple times in the target
+ * without recomputing it. For example, `f() ?: g()` should be translated to `if (f() != null) f() else g()`, but only one call
+ * to `f()` should be produced per usage of `f() ?: g()`. Note that not all sharing should be tread like this: for example, we share
+ * the condition node of a `while` loop, but it should be evaluated each time separately.
+ *
+ * This class contains the context that the sharing happens in, while the `sharedExp` is the result of the expression that is shared.
+ *
+ * This class should be used via the `share` function below. Do not do anything funny, or it will not work.
+ */
+data class SharingContext(override val inner: ExpEmbedding) : PassthroughExpEmbedding {
+    var sharedExp: Exp? = null
+
+    override fun <R> withPassthroughHook(ctx: LinearizationContext, action: LinearizationContext.() -> R): R =
+        ctx.action().also { sharedExp = null }
+
+    fun tryInitShared(f: () -> Exp): Exp = when (val r = sharedExp) {
```

Comment:
nit: in this case, you don't need to assign `sharedExp` to `r` since it is not used in the `null` case. And you can return `sharedExp` in the `else` branch.

extra-nit: WDYT about using `.also` here?

```kotlin
... = when (sharedExp) {
    null -> f().also { newSharedExp -> 
                sharedExp = newSharedExp
            }
    else -> sharedExp // Smart cast should work as expected here
}
```

Reply:
The compiler cannot guarantee that `sharedExp` is not modified between the two usages, so the assignment to a local is necessary.  I'll look into the `also`, thanks.

Reply:
> The compiler cannot guarantee that sharedExp is not modified between the two usages, so the assignment to a local is necessary.

Ah I see, thx!